### PR TITLE
Comfy

### PR DESCRIPTION
**Pull request by hlky**

### CODE REVIEW

**Comment by hlky** [comment]
@cruhl here you go, there's some code for "Extras" in the UI that I hadn't pushed yet too

Note that ComfyUI needs to be launched with `--enable-cors-header`

**Comment by cruhl** [comment]
@hlky Thanks so much! We'll likely start poking around on this Tuesday/Wednesday.

File: packages/stablestudio-plugin-comfyui/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,615 @@
+import * as StableStudio from "@stability/stablestudio-plugin";
+
+const basicWorkflow = (
```
Comment (hlky) [inline packages/stablestudio-plugin-comfyui/src/index.ts:3]: Text to image

File: packages/stablestudio-plugin-comfyui/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
+      class_type: "KSampler",
+      inputs: {
+        cfg: cfgScale,
+        denoise: 1,
+        latent_image: ["5", 0],
+        model: ["4", 0],
+        negative: ["7", 0],
+        positive: ["6", 0],
+        sampler_name: sampler,
+        scheduler: "normal",
```
Comment (hlky) [inline packages/stablestudio-plugin-comfyui/src/index.ts:25]: scheduler could be exposed as an extra

File: packages/stablestudio-plugin-comfyui/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,615 @@
...
+    "4": {
+      class_type: "CheckpointLoaderSimple",
+      inputs: {
+        ckpt_name: model,
+      },
+    },
+    "5": {
+      class_type: "EmptyLatentImage",
+      inputs: {
+        batch_size: 1,
```
Comment (hlky) [inline packages/stablestudio-plugin-comfyui/src/index.ts:39]: batch size could be an option

File: packages/stablestudio-plugin-comfyui/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,615 @@
...
+      class_type: "VAEDecode",
+      inputs: {
+        samples: ["3", 0],
+        vae: ["4", 2],
+      },
+    },
+    "9": {
+      class_type: "SaveImage",
+      inputs: {
+        filename_prefix: "ComfyUI",
```
Comment (hlky) [inline packages/stablestudio-plugin-comfyui/src/index.ts:68]: Filename prefix too

File: packages/stablestudio-plugin-comfyui/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,615 @@
...
+  negative_prompt: string,
+  model: string,
+  seed: number,
+  steps: number,
+  sampler: string,
+  cfgScale: number,
+  imageFilename: string,
+  imageStrength: number
+) => {
+  imageStrength = Math.max(0, Math.min(0.99, imageStrength));
```
Comment (hlky) [inline packages/stablestudio-plugin-comfyui/src/index.ts:86]: So denoise isn't 0 with 100% image strength

File: packages/stablestudio-plugin-comfyui/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,615 @@
...
+      class_type: "KSampler",
+      inputs: {
+        cfg: cfgScale,
+        denoise: 1 - imageStrength,
+        latent_image: ["11", 0],
+        model: ["4", 0],
+        negative: ["7", 0],
+        positive: ["6", 0],
+        sampler_name: sampler,
+        scheduler: "normal",
```
Comment (hlky) [inline packages/stablestudio-plugin-comfyui/src/index.ts:98]: This option and others could be exposed on all workflows

File: packages/stablestudio-plugin-comfyui/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,615 @@
...
+    "17": {
+      class_type: "VAEEncodeForInpaint",
+      inputs: {
+        grow_mask_by: 0,
+        mask: ["21", 0],
+        pixels: ["10", 0],
+        vae: ["4", 2],
+      },
+    },
+    "20": {
```
Comment (hlky) [inline packages/stablestudio-plugin-comfyui/src/index.ts:233]: This workflow worked with the mask sent by StableStudio when I tested this

File: packages/stablestudio-plugin-comfyui/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,615 @@
...
+    "21": {
+      class_type: "InvertMask",
+      inputs: {
+        mask: ["20", 0],
+      },
+    },
+  };
+};
+
+const getExtras = () => {
```
Comment (hlky) [inline packages/stablestudio-plugin-comfyui/src/index.ts:250]: Extras code isn't complete, the idea was to allow plugins to define extra UI elements

File: packages/stablestudio-plugin-comfyui/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,615 @@
...
+  extras.push(aitemplate);
+  extras.push(randomSlider);
+  return extras;
+};
+
+const getModels = async (apiUrl: string) => {
+  const response = await fetch(`${apiUrl}/object_info`);
+  if (response.ok) {
+    const json = await response.json();
+    return json.CheckpointLoaderSimple.input.required.ckpt_name[0];
```
Comment (hlky) [inline packages/stablestudio-plugin-comfyui/src/index.ts:295]: A similar request could be made to get specific model types from other `Loader's` input lists, VAE, LoRA etc

File: packages/stablestudio-plugin-comfyui/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,615 @@
...
+  } else {
+    return [];
+  }
+};
+
+const getSamplers = async (apiUrl: string) => {
+  const response = await fetch(`${apiUrl}/object_info`);
+  if (response.ok) {
+    const json = await response.json();
+    return json.KSampler.input.required.sampler_name[0];
```
Comment (hlky) [inline packages/stablestudio-plugin-comfyui/src/index.ts:305]: Schedulers should be available from KSampler input

File: packages/stablestudio-plugin-comfyui/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,615 @@
...
+      promptId: json.prompt_id,
+    };
+  } else {
+    return {
+      error: "Error",
+    };
+  }
+};
+
+const promptIdToImage = async (
```
Comment (hlky) [inline packages/stablestudio-plugin-comfyui/src/index.ts:347]: Returns images from history

File: packages/stablestudio-plugin-comfyui/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,615 @@
...
+      type: json.type,
+    };
+  } else {
+    return {
+      error: "Error",
+    };
+  }
+};
+
+const randomFilename = () => {
```
Comment (hlky) [inline packages/stablestudio-plugin-comfyui/src/index.ts:400]: This could be changed

File: packages/stablestudio-plugin-comfyui/src/index.ts [hunk 13]
```diff
@@ -0,0 +1,615 @@
...
+};
+
+const getDefaultInput = () => ({
+  width: 512,
+  height: 512,
+
+  cfgScale: 7,
+  steps: 30,
+  sampler: { id: "0", name: "euler" },
+  model: "0", //hack to select first model from list
```
Comment (hlky) [inline packages/stablestudio-plugin-comfyui/src/index.ts:418]: This should be looked at

File: packages/stablestudio-plugin-comfyui/src/index.ts [hunk 14]
```diff
@@ -0,0 +1,615 @@
...
+    };
+    input.model = model;
+    if (!input.model) {
+      return undefined;
+    }
+
+    const prompt =
+      input.prompts
+        ?.filter((p) => p.weight && p.weight > 0)
+        .map((p) => `(${p.text}:${p.weight})`)
```
Comment (hlky) [inline packages/stablestudio-plugin-comfyui/src/index.ts:475]: Wraps each input.prompt in (text:weight), there's only one prompt box but it should support the `Prompt strength`

File: packages/stablestudio-plugin-comfyui/src/index.ts [hunk 15]
```diff
@@ -0,0 +1,615 @@
...
+      width,
+      height,
+      seed,
+      steps,
+      sampler,
+      cfgScale
+    );
+    const initImageFilename = `${randomFilename()}.png`;
+    const maskImageFilename = `${randomFilename()}.png`;
+    if (input.initialImage) {
```
Comment (hlky) [inline packages/stablestudio-plugin-comfyui/src/index.ts:502]: If there's initialImage, upload it, and set workflow to image-to-image workflow

File: packages/stablestudio-plugin-comfyui/src/index.ts [hunk 16]
```diff
@@ -0,0 +1,615 @@
...
+        model,
+        seed,
+        steps,
+        sampler,
+        cfgScale,
+        initImageFilename,
+        input.initialImage.weight ?? 0.75
+      );
+    }
+    if (input.maskImage) {
```
Comment (hlky) [inline packages/stablestudio-plugin-comfyui/src/index.ts:523]: If there's maskImage the workflow is set to inpaint workflow

File: packages/stablestudio-plugin-comfyui/src/index.ts [hunk 17]
```diff
@@ -0,0 +1,615 @@
...
+        initImageFilename,
+        input.initialImage!.weight ?? 0.75,
+        maskImageFilename
+      );
+    }
+    let promptIds = [];
+    let images: StableStudio.StableDiffusionImages = { id: "", images: [] };
+    //messy, would be nicer to return images as they are ready
+    for (let i = 0; i < count; i++) {
+      workflow["3"].inputs.seed = seed + i;
```
Comment (hlky) [inline packages/stablestudio-plugin-comfyui/src/index.ts:549]: Increment seed, could be an option, random, fixed

File: packages/stablestudio-plugin/src/Plugin.ts [hunk 18]
```diff
@@ -215,6 +218,137 @@ export type Plugin<P extends PluginTypeHelper = PluginTypeHelperDefault> = {
   ) => void;
 } & P;
 
+export declare type UIExtraBadge = {
```
Comment (hlky) [inline packages/stablestudio-plugin/src/Plugin.ts:221]: Everything from `Theme` could be a UI extra

File: packages/stablestudio-ui/src/Editor/Dream/Render/index.tsx [hunk 19]
```diff
@@ -20,6 +20,9 @@ export namespace Render {
         entity,
         1
       );
+      //largest entity size
+      const width = Math.max(...entities.map((e) => e.width));
```
Comment (hlky) [inline packages/stablestudio-ui/src/Editor/Dream/Render/index.tsx:24]: Fixed outpainting iirc

File: packages/stablestudio-ui/src/Generation/Image/Sidebar/Extras.tsx [hunk 20]
```diff
@@ -0,0 +1,141 @@
+import { App } from "~/App";
+import { Generation } from "~/Generation";
+import { Theme } from "~/Theme";
+import { Plugin } from "~/Plugin";
+import * as StableStudio from "@stability/stablestudio-plugin";
+
+const extrasToElement = (extras: StableStudio.StableDiffusionExtra[]) => {
```
Comment (hlky) [inline packages/stablestudio-ui/src/Generation/Image/Sidebar/Extras.tsx:7]: As above, not complete or anything, but it maps the extras data to UI elements

**Review by hlky — COMMENTED** [review]
Should help get you up to speed with it

**Pull request closed without merging**